### ISSUE TITLE
Failure to Copy folder from one Storage Account to another

### ISSUE THREAD

**Issue by Imran-githu8** [microsoft/AzureStorageExplorer#8278]
### Preflight Checklist

- [x] I have installed the [latest version of Storage Explorer](https://github.com/Microsoft/AzureStorageExplorer/releases/latest).
- [x] I have checked existing resources, including the [troubleshooting guide](https://docs.microsoft.com/azure/storage/common/storage-explorer-troubleshooting) and the [release notes](https://github.com/microsoft/AzureStorageExplorer/releases).
- [x] I have searched for [similar issues](https://github.com/Microsoft/AzureStorageExplorer/issues).

### Storage Explorer Version

1.36.2

### Regression From

N/A

### Architecture

i86

### Storage Explorer Build Number

20241029.2

### Platform

Windows

### OS Version

Windows 11

### Bug Description

I am trying to copy a folder from one storage account to another storage account using Azure Storage Explorer and having the below:

```
{
  "name": null,
  "message": "\"failed to perform copy command due to error: cannot start job due to error: cannot list files due to reason GET https://xxxxxxxxxx.blob.core.windows.net/xxxxxxxxxx\\n--------------------------------------------------------------------------------\\nRESPONSE 403: 403 This request is not authorized to perform this operation.\\nERROR CODE: AuthorizationFailure\\n--------------------------------------------------------------------------------\\n﻿<?xml version=\\\"1.0\\\" encoding=\\\"utf-8\\\"?><Error><Code>AuthorizationFailure</Code><Message>This request is not authorized to perform this operation.\\nRequestId:30353ea5-f01e-0032-0b5f-3052ba000000\\nTime:2024-11-06T15:23:16.0031713Z</Message></Error>\\n--------------------------------------------------------------------------------\\n.\\n\""
}

```

I have added my Ip addresses in the network settings for both the storage accounts. And I am able to read and create files/folders perfectly fine through Azure Storage Explorer or the Azure Portal.

### Steps to Reproduce

1. Lauch Storage Explorer.
2. Expand to the Source location/Container in Storage account
3. Copy the Folder inside the Container.
4. Expand the Target Storage Account and Container inside it.
5. Paste the Folder inside the Container.

### Actual Experience

I am trying to copy a folder from one storage account to another storage account using Azure Storage Explorer and it is failing. 

I have added my Ip addresses in the network settings for both the storage accounts. And I am able to read and create files/folders perfectly fine through Azure Storage Explorer or the Azure Portal but copy is not working.,

### Expected Experience

It should copy the folder to another storage folder without any problems.

**Comment by craxal** [comment]
@Imran-githu8 A copy operation requires authentication for both the source and destination. Authentication for the source is accomplished by generating either a storage account key-based SAS or a user-delegated SAS. This is done for security reasons, because we don't want to paste user credentials onto the clipboard, and the valid time range of SAS can be controlled.

Do you have permission to list account keys or generate user-delegated SAS for the source storage account?

**Comment by Imran-githu8** [comment]
@craxal  yes I do have permissions to create SAS token for the storage account. 

Just to add I have tried with moving files and it works fine. But when I move folder with multiple folders and files inside it, then it fails.

**Comment by craxal** [comment]
@Imran-githu8 Are you working with ADLS Gen2 blob containers? If so, do you have the necessary permissions to read permissions to copy from the source items and the necessary write permissions to create the destination items?

@JasonYeMSFT Please add any additional comments if you have any.

**Comment by craxal** [comment]
Please also verify your machine architecture as well as the app architecture. We do not support i86 architectures.

**Comment by Imran-githu8** [comment]
@craxal  @JasonYeMSFT 

I checked the log files, and found the below but it didn't make much sense to me that why it is not able to check the destination file existence:

2024/11/21 10:59:39 ERR: [P#1-T#4118] COPYFAILED: https://xxxxxx.blob.core.windows.net/xxxxx/xxxx/DataFeedName=xxxxx/RunYear=xxxx/RunMonth=xx/RunDay=21/RunId=xxxxxxx/xxxxx-2657.json : 000 : Could not check destination file existence. context canceled
   Dst: https://xxxxxx.blob.core.windows.net/xxxx/xxxx/DataFeedName=xxxxx/RunYear=xxxx/RunMonth=xx/RunDay=21/RunId=xxxxxxxx/xxxxx-2657.json

**Comment by craxal** [comment]
What happens if you run the AzCopy command directly in the command line? You can click the "Copy AzCopy Command to Clipboard" action from Storage Explorer and paste it in the command line (might have to modify the command name to point to the correct location).

**Comment by Imran-githu8** [comment]
@craxal 

I have tried through power shell by installing azcopy on my machine and running the command:

"azcopy cp <source> <destination> --recursive"

but it gives me the same error that I am getting in Azure Storage Explorer:

![Image](https://github.com/user-attachments/assets/ef9cd0c1-3380-4e89-b284-fcf7c2f81e6f)

**Comment by Imran-githu8** [comment]
@craxal @JasonYeMSFT Do you have any idea what could be the issue?

I am feeling like I have exhausted all my options.

**Comment by MRayermannMSFT** [comment]
@Imran-githu8 can you share the SAS query parameters for your source and destingation, with the `sig` parameter redacted?

**Comment by Imran-githu8** [comment]
@MRayermannMSFT, thanks for helping, do you know how to extract that? is it from the azure portal under shared access signature:
![Image](https://github.com/user-attachments/assets/a2d25a47-ef14-46ec-8190-5eeb3849b218)

**Comment by craxal** [comment]
@Imran-githu8 It's the part of the string after the "?" For example:

?sv=2023-01-03&st=2024-12-03T19%3A54%3A39Z&se=2024-12-04T19%3A54%3A39Z&skoid=830a51ab-cfba-41c3-a07b-f69350bd0240&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2024-12-03T19%3A54%3A39Z&ske=2024-12-04T19%3A54%3A39Z&sks=b&skv=2023-01-03&sr=c&sp=rl&sig=REDACTED

**Comment by Imran-githu8** [comment]
thanks @craxal 

@MRayermannMSFT, 
**for source:**
?sp=r&st=2024-12-04T10:37:05Z&se=2024-12-04T18:37:05Z&spr=https&sv=2022-11-02&sr=c&sig=xYOFRkLQun2MYea******IfDVHyuGF3RgHI%2B4GCWmqc%3D

**for destination:**
?sp=r&st=2024-12-04T10:39:40Z&se=2024-12-04T18:39:40Z&spr=https&sv=2022-11-02&sr=c&sig=s6mwJH4UVSD2WLbaxCsk*******kDkwLnfa12lYD2GQ%3D

**Comment by Imran-githu8** [comment]
thanks @craxal

@MRayermannMSFT,
for source:
?sp=r&st=2024-12-04T10:37:05Z&se=2024-12-04T18:37:05Z&spr=https&sv=2022-11-02&sr=c&sig=******

for destination:
?sp=r&st=2024-12-04T10:39:40Z&se=2024-12-04T18:39:40Z&spr=https&sv=2022-11-02&sr=c&sig=*******

**Comment by MRayermannMSFT** [comment]
Thanks @Imran-githu8 ! I think the issue here are the permissions you have for your source and destination SAS tokens. I'm seeing them both as read-only. What you need is at least:

On the source:
- List: this allows Storage Explorer/AzCopy to enumerate the blobs you have in your container, thus discovering what needs to be copied
- Read: this allows Storage Explorer/AzCopy to read the contents of the blobs you have in your container

On the destination:
- List: this allows Storage Explorer/AzCopy to enumerate the blobs you have in your container, thus figuring out what already exists at the destination
- Write: this allows Storage Explorer/AzCopy to write to blobs you have in your container, thus actually copying over the data read from the source

How are you connecting to your containers?
- Are you connecting with SAS URLs/connection strings generated by you/someone else?
- Are you signing in, and then navigating to the container beneath a subscription?
- Are you signing in, and then connecting via OAuth and container URL?

**Comment by Imran-githu8** [comment]
@MRayermannMSFT  I am connecting using the sign in option in Azure Storage Explorer and then navigating to the container beneath a subscription, Just to add I am able to copy the files but when i copy the folder with multiple files it fails with the above error.

**Comment by craxal** [comment]
@Imran-githu8 What RBAC roles or permissions to do you have? Do you have read/write permissions?

**Comment by ghost** [comment]
I'm encountering this issue and my colleague as well, we are unable to copy a container to another. It copies some files and folders and suddenly cancels the whole copy with "Authentication failed, it is either not correct, or expired, or does not have the correct permission PUT"

Tried azcopy command with 2 SAS keys full of rights on the terminal as well...

EDIT: Solution was to edit the concurrency of the azcopy.
